code/modules/holomap/holomap.dm

```diff
@@ -0,0 +1,233 @@
+			var/area/A = get_area(T)
+			var/icon_to_use = 'icons/effects/holomap_parts.dmi'
+			var/icon_state_to_use
+			var/dir_to_use = 2
+			var/list/directions_to_aliens = list()
+
+			if(isfloorturf(T) && !locate(/obj/structure/window) in T)
+				icon_state_to_use = "floor"
+
+			else if(iswallturf(T) || istype(T, /turf/simulated/mineral) || locate(/obj/structure/window) in T)
+				var/list/cardinals = list(NORTH, SOUTH, EAST, WEST)
+				var/number_of_solid_turfs_neighbors = 0
+				var/space_found = FALSE
+				icon_state_to_use = "full"
+				for(var/i in 1 to 4)
+					var/turf/found_turf = get_step(T, cardinals[i])
+					var/area/found_area = get_area(found_turf)
+					if(found_area == /area/space || isspaceturf(found_turf))
+						icon_state_to_use = "full" //repeated since we might have set it to be "wall" already
+						space_found = TRUE
+						break
+					if(found_area != A || locate(/obj/machinery/door/airlock) in found_turf)
+						icon_state_to_use = "wall"
+						directions_to_aliens += cardinals[i] //this will work right... surely
+
+					if(iswallturf(found_turf) || istype(found_turf, /turf/simulated/mineral) || locate(/obj/structure/window) in found_turf)
+						number_of_solid_turfs_neighbors++
+
+				if(number_of_solid_turfs_neighbors >= 3)
+					icon_state_to_use = "full"
+					space_found = TRUE
+
+				if(!space_found)
+					switch(length(directions_to_aliens))
+						if(0)
+							icon_state_to_use = "full"
+						if(1)
+							dir_to_use = directions_to_aliens[1]
+						if(2)
+							if((directions_to_aliens[1] == NORTH && directions_to_aliens[2] == SOUTH) || (directions_to_aliens[1] == SOUTH && directions_to_aliens[2] == NORTH))
+								dir_to_use = NORTH
+							else if((directions_to_aliens[1] == EAST && directions_to_aliens[2] == WEST) || (directions_to_aliens[1] == WEST && directions_to_aliens[2] == EAST))
+								dir_to_use = EAST
+							else
+								if(directions_to_aliens[1] == NORTH)
+									if(directions_to_aliens[2] == EAST)
+										dir_to_use = NORTHEAST
+									else
+										dir_to_use = NORTHWEST
+								else if(directions_to_aliens[1] == SOUTH)
+									if(directions_to_aliens[2] == EAST)
+										dir_to_use = SOUTHEAST
+									else
+										dir_to_use = SOUTHWEST
+								else if(directions_to_aliens[1] == EAST )
+									if(directions_to_aliens[2] == NORTH)
+										dir_to_use = NORTHEAST
+									else
+										dir_to_use = SOUTHEAST
+								else
+									if(directions_to_aliens[2] == NORTH)
+										dir_to_use = NORTHWEST
+									else
+										dir_to_use = SOUTHWEST
+						if(3)
+							icon_state_to_use = "full"
+						if(4)
+							icon_state_to_use = "full"
+
+			else //if its space then nothing
+				continue
+
+			var/icon/I = icon(icon_to_use, icon_state_to_use, dir_to_use)
+
+			I.SwapColor(COLOR_WHITE, get_area_color(A))
+			I.SwapColor(COLOR_BLACK, get_area_color(A, TRUE))
+
+			bigassicon.Blend(I, ICON_OVERLAY, x_index*2, y_index*2)
```

Comment:
This is a really hefty operation. I'm not an expert, but this might need something like `CHECK_TICK` at the end